drivers/led/led_dac.c

```diff
@@ -0,0 +1,141 @@
+}
+
+static int led_dac_set_brightness(const struct device *dev, uint32_t led, uint8_t pct)
+{
+	const struct led_dac_config *config = dev->config;
+	uint32_t value;
+
+	if (led >= config->num_leds) {
+		return -EINVAL;
+	}
+
+	value = pct > 0 ? config->leds[led].dac_min + (uint64_t)(config->leds[led].dac_max -
+								 config->leds[led].dac_min) *
+							      pct / 100
+			: 0;
+
+	return led_dac_set_raw(dev, led, value);
+}
+
+static inline int led_dac_on(const struct device *dev, uint32_t led)
+{
+	const struct led_dac_config *config = dev->config;
+
+	if (led >= config->num_leds) {
+		return -EINVAL;
+	}
+
+	return led_dac_set_raw(dev, led, config->leds[led].dac_max);
+}
+
+static inline int led_dac_off(const struct device *dev, uint32_t led)
+{
+	const struct led_dac_config *config = dev->config;
+
+	if (led >= config->num_leds) {
+		return -EINVAL;
+	}
+
+	return led_dac_set_raw(dev, led, 0);
+}
+
+static DEVICE_API(led, led_dac_api) = {
+	.on = led_dac_on,
+	.off = led_dac_off,
+	.set_brightness = led_dac_set_brightness,
+};
+
+static int led_dac_init(const struct device *dev)
+{
+	const struct led_dac_config *config = dev->config;
+	int ret;
+
+	for (uint8_t i = 0; i < config->num_leds; ++i) {
+		const struct led_dac_leds *led = &config->leds[i];
+
+		if (!device_is_ready(led->dac)) {
+			return -ENODEV;
+		}
+
+		ret = dac_channel_setup(led->dac, &led->chan_cfg);
+		if (ret != 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+#define LED_DAC_MAX_MV(n)  DT_PROP(n, voltage_max_dac_mv)
+#define LED_DAC_MAX_VAL(n) (BIT(DT_PROP(n, resolution)) - 1)
+
+#define LED_DAC_MAX_BRIGHTNESS(n)                                                                  \
+	COND_CODE_1(DT_NODE_HAS_PROP(n, voltage_max_brightness_mv),                                \
+	     (DT_PROP(n, voltage_max_brightness_mv) * LED_DAC_MAX_VAL(n) / LED_DAC_MAX_MV(n)),     \
+	     (LED_DAC_MAX_VAL(n)))
+
+#define LED_DAC_MIN_BRIGHTNESS(n)                                                                  \
+	COND_CODE_1(DT_NODE_HAS_PROP(n, voltage_min_brightness_mv),                                \
+	     (DT_PROP(n, voltage_min_brightness_mv) * LED_DAC_MAX_VAL(n) / LED_DAC_MAX_MV(n)),     \
+	     (0))
+
+#define LED_DAC_DT_GET(n)                                                                          \
+	{                                                                                          \
+		.dac = DEVICE_DT_GET(DT_PHANDLE(n, dac_dev)),                                      \
+		.chan_cfg = {.channel_id = DT_PROP(n, channel),                                    \
+			     .resolution = DT_PROP(n, resolution),                                 \
+			     .buffered = DT_PROP(n, output_buffer),                                \
+			     .internal = false},                                                   \
+		.dac_max = LED_DAC_MAX_BRIGHTNESS(n), .dac_min = LED_DAC_MIN_BRIGHTNESS(n)         \
```

Comment:
Suggested format to improve readibility (the `\` characters still need to be re-aligned):
```suggestion
		.chan_cfg = {
			.channel_id = DT_PROP(n, channel),  \
			.resolution = DT_PROP(n, resolution), \
			.buffered = DT_PROP(n, output_buffer), \
			.internal = false, \
		}, \
		.dac_max = LED_DAC_MAX_BRIGHTNESS(n), \
		.dac_min = LED_DAC_MIN_BRIGHTNESS(n), \
```

Reply:
`clang-format` is not always good friends with macros. But appending brightness to the variables make line length require newline :+1:

Reply:
Yeah, and you can add a comma to the last element to enforce that every element is put into a new line and the braces are set more nicely.